broadinstitute/single_cell_portal_core
# Adding `AnnDataFileInfo` and `AnnDataIngestParameters` to manage `AnnData` ingest jobs (SCP-4831)

#### BACKGROUND
Since `AnnData` files can contain many data modalities - clustering, metadata, expression - we need a way of tracking what data has been extracted from an `AnnData` file.  Furthermore, we will need a mechanism for setting the various parameters for downstream ingest jobs once SCP file types have been extracted from a parent `AnnData` file.

#### CHANGES
This update adds two new models: `AnnDataFileInfo` and `AnnDataIngestParameters`.  `AnnDataFileInfo` is a small embedded class similar to `ExpressionFileInfo` that will track what data has been extracted from an `AnnData` file.  It also provides an interface for using the normal `Study`-bound methods for finding files of a given type, such as `study.expression_matrices`.  This ensures `AnnData` files that have had the corresponding data type extracted will be identified correctly.

`AnnDataIngestParameters` is analogous to the other `Parameterizable` classes that handle setting command line parameters for PAPI jobs, such as `DifferentialExpressionParameters`.  This will allow SCP to manage passing information from one job to the next without having to pull pipeline job metadata objects and parsing out command lines to get values for certain parameters.  Eventually, `AnnDataIngestParameters` will be able to handle all ingest-specific actions for `AnnData` files, from the initial "fragment" SCP file extracting to each downstream parse job for clustering, metadata, and expression.  Currently, it can handle extracting clustering files from original `h5ad` archives and launching `ingest_cluster` jobs once the extraction is complete.

#### MANUAL TESTING
1. Download https://cellxgene.cziscience.com/collections/2902f08c-f83c-470e-a541-e463e25e5058
2. Pull branch and boot all services to be able to upload and parse files (e.g. `bin/docker-compose-setup.sh`)
3. Sign into your local instance, go to the "Admin" interface and set the `Ingest Pipeline Docker Image` to `gcr.io/broad-singlecellportal-staging/scp-ingest-jb-anndata-delocalize-bugfix:896b506`
4. Go to the "Feature Flags" interface, and set the `ingest_anndata_file` flag for your email to `true`
5. Create a new study, and upload `local.h5ad` as an `AnnData` file, and confirm that once the file uploads a job is launch to perform the initial extraction
6. After 1-2 minutes, you should receive an email saying that the `AnnData file ingest has completed`
7. In `development.log`, you should see two new jobs launched to ingest the extracted clustering files for `X_tsne` and `X_umap`:
```
Remote found for local.h5ad:638e2ea4ace7570f36a9ff0e, launching Ingest job
2022-12-05 12:53:14 -0500: Deleting caches for SCP81
2022-12-05 12:53:14 -0500: caches successfully cleared
Request object sent to Google Pipelines API (PAPI), excluding 'environment' parameters:
---
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 63878fb7d17c6300547fe22b
  - "--study-file-id"
  - 638e2ea4ace7570f36a9ff0e
  - "--user-metrics-uuid"
  - f775823b-94f7-46af-b1c7-789582efaa18
  - ingest_cluster
  - "--ingest-cluster"
  - "--cluster-file"
  - gs://fc-45f2ff6d-92f5-4cad-9c6b-a4a832c2a2d7/_scp_internal/anndata_ingest/X_tsne.cluster.anndata_segment.tsv
  - "--name"
  - X_tsne
  - "--domain-ranges"
  - "{}"
  :flags: []
  :image_uri: gcr.io/broad-singlecellportal-staging/scp-ingest-jb-anndata-delocalize-bugfix:896b506
  :labels: {}
  :timeout: {}
```
8. After another 1-2 minutes, you should receive success emails for the two extracted clustering files
9. Go to the "Details" page for your study, and confirm that you see two new cluster objects with the following info:
![Screen Shot 2022-12-05 at 1 29 24 PM](https://user-images.githubusercontent.com/729968/205714237-157f77fa-2d5a-445c-bd22-1501d13f6405.png)
